packages/frontend/core/src/components/workspace-upgrade/upgrade-hooks.ts

```diff
@@ -1,42 +1,126 @@
+  const rootStore = getCurrentStore();
 
-  const upgradeWorkspace = useCallback(() => {
+  const upgradeWorkspace = useAsyncCallback(async () => {
     setState('upgrading');
     setError(null);
+    try {
+      if (environment.isDesktop) {
+        throw new Error('Migration should happened in client starting stage.');
+      }
 
-    (async () => {
+      // Migration need to wait for root doc and all subdocs loaded.
       await syncEngine?.waitForSynced();
-      await forceUpgradePages({
-        getCurrentRootDoc: async () => workspace.blockSuiteWorkspace.doc,
-        getSchema: () => workspace.blockSuiteWorkspace.schema,
+
+      // Clone a new doc to prevent change events.
+      const clonedDoc = new YDoc({
+        guid: workspace.blockSuiteWorkspace.doc.guid,
+      });
+      applyDoc(clonedDoc, workspace.blockSuiteWorkspace.doc);
+      const schema = workspace.blockSuiteWorkspace.schema;
+      let newWorkspace: Workspace | null = null;
+
+      const resultDoc = await migrateWorkspace(migration, {
+        doc: clonedDoc,
+        schema,
+        createWorkspace: () => {
+          // Migrate to subdoc version need to create a new workspace.
+          // It will only happened for old local workspace.
+          newWorkspace = getOrCreateWorkspace(nanoid(), WorkspaceFlavour.LOCAL);
+          return Promise.resolve(newWorkspace);
+        },
       });
 
+      if (newWorkspace) {
+        const localMetaString =
+          localStorage.getItem('jotai-workspaces') ?? '[]';
+        const localMetadataList = JSON.parse(
+          localMetaString
+        ) as RootWorkspaceMetadata[];
+        const currentLocalMetadata = localMetadataList.find(
+          item => item.id === workspace.id
+        );
+        const flavour = currentLocalMetadata?.flavour ?? WorkspaceFlavour.LOCAL;
+
+        // Legacy logic moved from `setup.ts`.
+        // It works well before, should be refactor or remove in the future.
+        const adapter = WorkspaceAdapters[flavour];
+        const newId = await adapter.CRUD.create(newWorkspace);
+        const [workspaceAtom] = getBlockSuiteWorkspaceAtom(newId);
+        await rootStore.get(workspaceAtom); // Trigger provider sync to persist data.
+
+        await adapter.CRUD.delete(workspace.blockSuiteWorkspace);
+        await migrateLocalBlobStorage(workspace.id, newId);
+        setNewWorkspaceId(newId);
+
+        const index = localMetadataList.findIndex(
+          meta => meta.id === workspace.id
+        );
+        localMetadataList[index] = {
+          ...currentLocalMetadata,
+          id: newId,
+          flavour,
+        };
+        localStorage.setItem(
+          'jotai-workspaces',
+          JSON.stringify(localMetadataList)
+        );
+        localStorage.setItem('last_workspace_id', newId);
+        localStorage.removeItem('last_page_id');
+      } else {
+        applyDoc(workspace.blockSuiteWorkspace.doc, resultDoc);
+      }
+
       await syncEngine?.waitForSynced();
```

Comment:
this `waitForSynced` is called after migration is to wait for changes to be written to data sources?

Reply:
We will display a "refresh page" button. After this sync is completed, the user can safely click refresh.